batch ucrpillow

sravfeyn opened the pull request:
This PR
1. Introduces a `BulkPillowProcessor` class
2. Extends the current `PillowBase` to pass changes in chunks if the attached processor is `BulkPillowProcessor` (and couple of other conditions, see code for details)
3. Implements `process_change_chunk` for the UCR `ConfigurableReportKafkaPillow`

sravfeyn committed:
add chunked processing support for pillow

sravfeyn committed:
simplify chunk spec

sravfeyn committed:
validation; tests; make ucr pillow fake chunked

sravfeyn committed:
create BulkPillowProcessor

sravfeyn committed:
remove print

sravfeyn committed:
unused

sravfeyn committed:
merge master

sravfeyn committed:
stricter check

sravfeyn committed:
add to init

sravfeyn commented on corehq/ex-submodules/pillowtop/processors/__init__.py:
```diff
@@ -1,3 +1,3 @@
-from .interface import PillowProcessor
+from .interface import PillowProcessor, BulkPillowProcessor
```
Ignoring since this is init module

sravfeyn committed:
Merge branch 'master' into sr-batch-ucrpillow

sravfeyn committed:
pass doc

sravfeyn committed:
implement adapter.bulk_save

sravfeyn committed:
implement bulk_update_records

sravfeyn committed:
implement ucr bulk-pillow-processing

sravfeyn committed:
Merge branch 'sr-exp1' into sr-batch-ucrpillow

sravfeyn committed:
move

sravfeyn committed:
todo

sravfeyn commented:
@emord @snopoke  There are couple of small todos/edge-cases but most of the code is in place now. So it will be good to get a review now

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,28 +100,74 @@ def run(self):
+        if not changes_chunk:
+            return
+        try:
+            # chunked processing is supported if there is only one processor
```
This constraint is kept to keep things simple and is not a very hard constraint

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,28 +100,74 @@ def run(self):
+        try:
+            # chunked processing is supported if there is only one processor
+            result = self.processors[0].process_changes_chunk(self, changes_chunk)
+        except BulkPorcessingError as ex:
```
If the bulk processor raises an error that is not related to any particular change, the pillow will just fall back to serial processing.

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,28 +100,74 @@ def run(self):
+        else:
+            for change in result.succeeded_changes:
+                self._update_checkpoint(change, context)
+                self._record_change_success_in_datadog(change)
```
Is it necessary to call this for every change? Can't we just do a single call with the number of changes?

sravfeyn replied:
Yeah, we should, I am planning to change that. I haven't planned dd-metrics for bulk processing yet

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -222,6 +225,81 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                except Exception as e:
+                    raise BulkPorcessingError('Error bulk saving calculated UCR rows {}'.format(e))
+
+            succeeded_changes.union(changes_chunk)
```
Some of these could be failures. This should only include changes for docs that were returned by `get_docs`

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -133,23 +133,39 @@ def get_distinct_values(self, column, limit):
+        rows = []
+        for doc in docs:
+            rows.extend(self.get_all_values(doc))
+        self.save_rows(rows)
```
missing error handling

sravfeyn replied:
Addressed in other PR. Planning to add error-handling

snopoke commented on corehq/apps/userreports/tests/utils.py:
```diff
@@ -146,3 +152,40 @@ def get_simple_xform():
     return xform.tostring()
+
+
+def load_data_from_db(table_name):
```
this is only used in one place and it only checks the length of the results. Seems like a lot of unnecessary code.

sravfeyn replied:
Addressed in other PR

snopoke reviewed (commented):
Didn't do a very thorough review but made a few comments

sravfeyn committed:
merge master

sravfeyn committed:
implement default bulk-save

sravfeyn committed:
implement bulk delete

sravfeyn committed:
remove sql specific bulk-save

sravfeyn committed:
pass full doctypes

sravfeyn committed:
support multiple doctypes, error handling, result

sravfeyn committed:
imports

sravfeyn committed:
tests for chunked processor

sravfeyn committed:
tests

sravfeyn committed:
import

sravfeyn commented on corehq/apps/userreports/models.py:
```diff
@@ -883,6 +883,29 @@ def bulk_creation(cls, doc_ids, doc_type, domain, config_ids):
+
+        old_indicators = AsyncIndicator.objects.filter(doc_id__in=doc_ids).all()
+        to_update = []
+        from bulk_update.helper import bulk_update as bulk_update_helper
```
Updated.

sravfeyn commented on corehq/apps/userreports/pillow.py:
```diff
@@ -222,6 +225,101 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                if change.id not in docs_by_id:
+                    # we need to capture DocumentMissingError which is not possible in bulk
+                    #   so let pillow fall back to serial mode to capture the error for missing docs
+                    failed_changes.append(change)
```
Sure

sravfeyn committed:
continue

sravfeyn committed:
datadog for chunked processing

sravfeyn committed:
metrics

sravfeyn commented:
@emord This is now ready to be merged (except for two quick comments)

sravfeyn committed:
correction

sravfeyn committed:
merge master

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +100,94 @@ def run(self):
+        else:
+            succeeded_changes = set(changes_chunk) - set(failed_changes)
+            for change in succeeded_changes:
+                self._update_checkpoint(change, context)
```
Is it okay to update checkpoints in this manner in chunked processing?

snopoke replied:
My comment on this seems to been misplaced:

I haven't thought about this much but does this checkpointing process still work when doing in bulk? For couch it will just take the sequence from the change object and for kafka it will take it from the change feed. Certainly for Kafka it doesn't seem to make sense to call this for every successful change since the offset of the change feed will be the same.

There's also the question of whether we should be recording the current offset or the max offset before we hit an error. I think what we want is to make sure we don't set the checkpoint to any value beyond which we've successfully processed a change (or successfully handled the error).

sravfeyn committed:
track timer

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +100,94 @@ def run(self):
+                self.process_with_error_handling(change, context, chunked_fallback=True)
+        self._record_datadog_metrics(changes_chunk, failed_changes)
+
+    def _record_datadog_metrics(self, changes_chunk, failed_changes):
```
The one that's missing here is a slow-ucr-lag metric similar to the [one](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/pillow.py#L278) for serial processor. I am not sure how that is defined to be able to define the same here. Do you have any ideas?

sravfeyn replied:
> I don't think that they need to be namespaced with "chunked"

I will move the chunked to a tag.

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -222,6 +225,101 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+            # query
+            docs = []
+            for doc_store, d_changes in six.iteritems(_changes_by_doc_store):
+                docs.extend(list(doc_store.iter_documents([change.id for change in d_changes])))
```
I think this is fine for doc sizes of 100, but if we ever went larger we probably wouldn't want it as a list

sravfeyn replied:
Do you think it would be anymore orders of magnitude?

emord commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +100,94 @@ def run(self):
+        max_change_lag = (datetime.utcnow() - changes_chunk[0].metadata.publish_timestamp).seconds
+        min_change_lag = (datetime.utcnow() - changes_chunk[-1].metadata.publish_timestamp).seconds
+        perceived_lag = (max_change_lag - min_change_lag) / 2
+        datadog_gauge('commcare.change_feed.chunked.perceived_change_lag', perceived_lag, tags=[
```
What does `perceived_change_lag` mean? Is it the amount of time we've been collecting the chunks for?

Similarly I don't quite understand `actual_change_lag` It looks like it's the average of all documents lags, which I'm not sure how it would be used.

How do you think  these will be used?

sravfeyn replied:
I will make this simple and keep just min/max change-lags. 

Let's say we have 100 changes, with the earliest change 2 seconds ago and last change 1 second ago. If we take 10sec to process these 100 changes, the min_change_lag would be 10sec - 2sec, and max would be 10sec - 1sec.

Also interesting, that in the chunked processing the change lag might actually go up, since we wait to process changes in a minimum chunk. But overall because we process them in chunks, the processing time per change should go down.

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -222,6 +225,101 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+
+        return failed_changes
+
     def process_change(self, pillow_instance, change):
```
`process_change` isn't used anymore right? could that be removed?

sravfeyn replied:
It's used as a fall back. If there is an exception in `process_changes_chunk`, the pillow will fall back to using `process_change` one by one. 
Additionally, I think it might be useful to keep both to toggle chunked mode for a pillow

emord commented on corehq/apps/userreports/models.py:
```diff
@@ -883,6 +884,29 @@ def bulk_creation(cls, doc_ids, doc_type, domain, config_ids):
+        to_update = []
+
+        for indicator in old_indicators:
+            if set(indicator.indicator_config_ids) != set(configs_by_docs[indicator.doc_id]):
```
Instead of `!=` I think that we want to see if the new config_ids are a subset of the old config_ids. If it is a subset, then we don't need to update. 

If they don't match we want the superset of both the old config ids and new config ids

sravfeyn replied:
Ha, I see. Will update

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +100,94 @@ def run(self):
-                    self.process_with_error_handling(change, context)
+                    if self._should_process_in_chunks:
+                        changes_chunk.append(change)
+                        if len(changes_chunk) == self.processors[0].processor_chunk_size:
```
I thought that the tmeout handling is taken care of in `self.get_change_feed().iter_changes`, isn't it so?

snopoke replied:
@emord means a timeout for setting the checkpoint. You can see how we currently handle that here: https://github.com/dimagi/commcare-hq/blob/232b6696412e4fbbd9df4b05f55b8738af7efe33/corehq/ex-submodules/pillowtop/checkpoints/manager.py#L133

sravfeyn replied:
The manager example deals with current iteration and previous iteration (so snapshots are available to compare). Here we need to deal with current iteration and an expected next iteration that might be delayed, so I don't see an easy way without using some sort of parallel polling.

The only graceful way of doing this might be to have  `self.get_change_feed().iter_changes` re-raise `ConsumerTimeout` error and processing remaining chunk and restart `process_changes` (similar to how `PillowCheckpointReset` is handled currently. 

Is there a better way?

snopoke replied:
You'll have keep track of the last time the processing was called:
```
if len(changes_chunk) >= self.processors[0].processor_chunk_size or datetime.utcnow() - last_process_time > timedelta(minutes=X):
     last_process_time = datetime.utcnow()
     ....
```

sravfeyn replied:
This works only if the next change comes after a certain `MAX_WAIT_TIME`, since this condition  gets evaluated only if there is a next change. Better than having nothing, but still won't guard us waiting for a long time with an unprocessed chunk. I will add this logic with a very short `MAX_WAIT_TIME` like 10 seconds, let me know if otherwise.

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +100,94 @@ def run(self):
+    def _record_datadog_metrics(self, changes_chunk, failed_changes):
+        tags = ["pillow_name:{}".format(self.get_name())]
+        datadog_counter('commcare.change_feed.chunked.changes', len(changes_chunk), tags=tags)
+        datadog_counter('commcare.change_feed.chunked.changes.exception', len(changes_chunk), tags=tags)
```
Yeah, my mistake, will update

sravfeyn commented on corehq/ex-submodules/pillowtop/processors/sample.py:
```diff
@@ -47,3 +47,17 @@ def __init__(self):
         self.changes_seen.append(change)
+
+
+class ChunkedCountProcessor(BulkPillowProcessor):
```
There are more [here](https://github.com/dimagi/commcare-hq/pull/20834/files#diff-67129b077cd07e96bf8f4f4ffdd44b55R64). This only unit-tests the pillow methods, not the processor methods

sravfeyn replied:
We could move, but it's named `sample.py` anyway

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -222,6 +225,101 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                    failed_changes.union(to_update)
+
+            if async_configs_by_doc_id:
+                # Todo; handle multiple doc-types
```
is this todo still relevant?

sravfeyn replied:
Nope, will remove it

sravfeyn committed:
change update criteria

sravfeyn committed:
1000 -> 100

sravfeyn committed:
checkpoint update strategy for bulk changes

sravfeyn commented on corehq/apps/userreports/pillow.py:
```diff
@@ -352,6 +453,7 @@ def get_kafka_ucr_static_pillow(pillow_id='kafka-ucr-static', ucr_division=None,
             ucr_division=ucr_division,
             include_ucrs=include_ucrs,
             exclude_ucrs=exclude_ucrs,
+            processor_chunk_size=processor_chunk_size,
```
Updated

sravfeyn committed:
modify datadog metrics

sravfeyn committed:
fix tags and error

sravfeyn commented:
Will fix the conflicts once this [PR](https://github.com/dimagi/commcare-hq/pull/20964) gets in

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +101,101 @@ def run(self):
 
+    @property
+    def _should_process_in_chunks(self):
+        # chunked processing is only supported for pillows with single processor
```
Why is this necessary?

sravfeyn replied:
It's not a hard requirement. I kept this constraint to keep the `process_changes` logic (pertaining to checkpoint updates and multiple processors having different `processor_chunk_size`) simple and also considering that there is only one pillow that has multiple processors right now (and none of them a bulk processor). It's not difficult to get rid of this constraint, but it wasn't necessary when I started the PR

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -146,30 +144,34 @@ def process_chunk_with_error_handling(self, changes_chunk, context):
                 for change in failed_changes:
                     self.process_with_error_handling(change, context, chunked_fallback=True)
+        # rewind `changes_seen` counter to update checkpoint one by one
+        context.changes_seen = context.changes_seen - len(changes_chunk)
```
I thought it wasn't possible to avoid this, but updated to remove this

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -114,24 +114,22 @@ def _should_process_in_chunks(self):
         try:
             changes_chunk = []
             for change in self.get_change_feed().iter_changes(since=since or None, forever=forever):
-                if change:
```
Mainly because we need to skip `self._update_checkpoint(None, None)` if it's a empty change and if it's the processor is bulk. Also, it simplifies this block

sravfeyn replied:
Actually, reverting that change

sravfeyn committed:
remove back n forth

sravfeyn committed:
Refecator checkpoint update

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -146,30 +144,34 @@ def process_chunk_with_error_handling(self, changes_chunk, context):
+        context.changes_seen = context.changes_seen - len(changes_chunk)
+        for change in changes_chunk:
+            context.changes_seen += 1
+            self._update_checkpoint(change, context)
```
Yeah, I think we can do this just for the very last change actually. The current code already checks this condition and doesn't update for every change even if we ask it to unless the change count frequency or timehit matches. (See `should_update_checkpoint`).

Also refactored some of the update-checkpoint functions. https://github.com/dimagi/commcare-hq/pull/20834/commits/a79ef37bc2272ec34e0368fd86c0c7687f31b2b9

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -219,7 +219,7 @@ def process_change(self, change):
         pass
 
     @abstractmethod
-    def fire_change_processed_event(self, change, context):
```
Got rid of most of `fire_change_processed` methods and renamed/refactored it into `update_checkpoint`, since that's the only thing that `fire_change_processed` was doing

sravfeyn committed:
dont update checkpoint for every change

sravfeyn committed:
guard against slow change feed

emord commented:
I got really confused when trying to review this. I think that since the batch async indicator PR was reverted this PR is containing code  that was originally in that PR. Will take another look once merge conflicts are resolved

sravfeyn committed:
PR feedback

emord reviewed (commented):
Looks like htere are some merge conflicts due to ES, but it looks good

sravfeyn commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +101,103 @@ def run(self):
+                for change in failed_changes:
+                    self.process_with_error_handling(change, context, chunked_fallback=True)
+        for change in changes_chunk:
+            context.changes_seen += 1
```
Sure, probably not that a big deal either way.

sravfeyn committed:
small change

sravfeyn committed:
merge master; resolve conflicts

sravfeyn committed:
resolve conflicts

sravfeyn committed:
correct type

sravfeyn commented:
Bump.

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                    notify_exception(None,
+                        "{name} Error in saving changes chunk {ids}: {ex}".format(
+                            name=pillow_instance.get_name(), ids=[c.id for c in to_delete], ex=ex))
+                    failed_changes.union(to_update)
```
this just marks the all as failed including the async ones. would be nice if we could make it more specific

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -58,6 +58,94 @@ def test_needs_bootstrap_window(self):
+        docstore_patch.return_value = docs[0:5]
+        cases = self._create_and_process_changes(docs=docs)
+        query = self.adapter.get_query_object()
+        # first five docs should be processed in bulk, last five serially
```
could also patch `process_change` to verify that it got called 5 times

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -187,9 +293,18 @@ def test_process_doc_from_sql(self, datetime_mock):
+    @mock.patch('corehq.apps.userreports.specs.datetime')
+    def test_process_deleted_doc_from_sql_chunked(self, datetime_mock):
+        self.pillow = get_kafka_ucr_pillow(processor_chunk_size=100)
+        self._test_process_deleted_doc_from_sql(datetime_mock)
```
It would be cleaner to pass in the pillow rather that update the reference on `self`. That would also prevent interference with other tests if one of these failed.

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -358,6 +481,16 @@ def _test_pillow(self, pillow, since):
+    def test_reuse_cache_chunked(self):
+        self.pillow1 = get_kafka_ucr_pillow(topics=['case-sql'], processor_chunk_size=100)
+        self.pillow2 = get_kafka_ucr_pillow(topics=['case-sql'], processor_chunk_size=100)
+        self._test_reuse_cache()
```
similar comment here about passing in the pillows

snopoke reviewed (commented):
Did a pass but I'd like to take another look before merging. Sorry to delay.

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
             # remove it until the next bootstrap call
             self.table_adapters_by_domain[domain].remove(table)
 
+    def process_changes_chunk(self, pillow_instance, changes):
```
Can this method be broken up into helpers? It's getting a bit hard to follow at ~100 lines including the internal function. I think it could be written as:


```python
def process_changes:
    relevant_changes = sorted([change for change in changes if change.domain in self.domains_with_adapters])
    changes_by_domain = # {domain: [change1,change2, ...], ...}
    for domain in changes_by_domain:
        self.process_changes_by_domain(domain, changes_by_domain[domain])

def process_changes_by_domain(domain, changes):
    for changes in changes_by_doc_store:
        self.process_changes_by_doc_store(doc_store, changes)
```

(I think that you're already processing them by domain/doc store, but I think this will make it a little clearer.)

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            rows_to_save_by_adapter = defaultdict(list)
+            adapters = list(self.table_adapters_by_domain[domain])
```
Is the coercion to `list` necessary?

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                        else:
+                            try:
+                                rows_to_save_by_adapter[adapter].extend(adapter.get_all_values(doc, eval_context))
+                            except Exception as ex:
```
These exceptions are very general/catch all. Is it possible to make the more specific to what you expect could may be failing?

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            rows_to_save_by_adapter = defaultdict(list)
+            adapters = list(self.table_adapters_by_domain[domain])
+            to_delete = {change for change in changes_chunk if change.deleted}
```
I think that to_delete only ever needs the ID. so this could be changed to `{change.id for change in changes_chunk if change.deleted}` which would remove some list comprehensions later in the code

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -273,13 +370,14 @@ class ConfigurableReportKafkaPillow(ConstructedPillow):
 
-    def __init__(self, processor, pillow_name, topics, num_processes, process_num, retry_errors=False):
+    def __init__(self, processor, pillow_name, topics, num_processes, process_num, retry_errors=False,
+                checkpoint_frequency=1000):
```
looks like `checkpoint_frequency` is always passed in and has a value of 100, which is confusing when default arg is different from the most common passed value

emord commented on corehq/apps/userreports/adapter.py:
```diff
@@ -82,8 +82,10 @@ def bulk_save(self, docs):
         for doc in docs:
-            self.save(doc)
+            rows.extend(self.get_all_values(doc))
+        self.save_rows(rows)
```
I think this should still be on the `SqlAdapter` class as `save_rows` isn't defined on this base class, so this would error if we implemented a different database again.

sravfeyn committed:
Merge branch 'master' into sr-batch-ucrpillow

sravfeyn committed:
move methods between inheritance classes

sravfeyn committed:
review feedback

sravfeyn committed:
break large function

sravfeyn committed:
naming, correct var

sravfeyn committed:
some of docs are to be deleted per adapter

sravfeyn committed:
let adapter handle the exception

sravfeyn committed:
simplify vars

sravfeyn committed:
terminology

sravfeyn committed:
update kwargs

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,117 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                table.handle_exception(doc, e)
+            except UserReportsWarning:
+                # remove it until the next bootstrap call
+                self.table_adapters_by_domain[domain].remove(table)
```
removing the table from the list might cause other tables to be skipped down on line 297:
```
l = [1,2,3]
for i in l:
   print(i)
   l.remove(i)
>> 1
>> 3
```

emord replied:
I think that we don't want to remove these at all, as it could cause missing documents in the data source.

1. Change a column type, and rebuild it
2. A change comes in and errors because the data source was kept in memory and doesn't match the database definiton
3. The data source is removed here
4. Another change comes in, but there is no data source so it's not processed/no error is emitted.
5. that change never gets to the data source

emord replied:
Thinking of this now, it's possible that we're open to a type of this errors now as pillows have no way to be notified that a new data source has been created. This could have been the source of bugs before.

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +211,117 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+
+        for doc in docs:
+            eval_context = EvaluationContext(doc)
+            for adapter in adapters:
```
You'll need to make a copy of this list in order to avoid the issue I described above about skipping items in the list

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +101,102 @@ def run(self):
+        self._record_datadog_metrics(changes_chunk, retry_changes, timer)
+
+    def _record_datadog_metrics(self, changes_chunk, retry_changes, timer):
+        tags = ["pillow_name:{}".format(self.get_name()), "chunked:True"]
```
instead of `chunked:True` I'd use `mode:chunked`

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +101,102 @@ def run(self):
+
+        max_change_lag = (datetime.utcnow() - changes_chunk[0].metadata.publish_timestamp).seconds
+        min_change_lag = (datetime.utcnow() - changes_chunk[-1].metadata.publish_timestamp).seconds
+        datadog_gauge('commcare.change_feed.chunked.min_change_lag', min_change_lag, tags=tags)
```
I don't think min_lag is really of much use.

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,29 +101,102 @@ def run(self):
+        datadog_histogram(
+            'commcare.change_feed.processing_time',
+            timer.duration / len(changes_chunk),
+            tags=tags + ["chunk_size:".format(str(len(changes_chunk)))])
```
don't think we need the `chunk_size` tag (also there's no format parameter)

sravfeyn committed:
hash correctly, unit test get docs and more coverage

sravfeyn committed:
be careful when removing from loop source

sravfeyn committed:
move code

sravfeyn commented:
@emord @snopoke thanks for the review. I have addressed it. If it's good to go, please don't merge, I would like to test this on staging first.

sravfeyn committed:
update tags

sravfeyn committed:
kwarg

sravfeyn committed:
remove single processor constraint for bulk processors

sravfeyn committed:
Fix test

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -48,6 +48,8 @@ class PillowBase(six.with_metaclass(ABCMeta, object)):
     # set to true to disable saving pillow retry errors
     retry_errors = True
+    # this will be the batch size for processors that support batch processing
+    processor_chunk_size = 0
```
might make more sense to default this to `None`

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -109,10 +111,18 @@ def _update_checkpoint(self, change, context):
-        return (len(self.processors) == 1
-            and getattr(self.processors[0], 'processor_chunk_size', 0) > 0)
+    def batch_processors(self):
+        if self.processor_chunk_size > 0:
```
nit: could be just `if self.processor_chunk_size`

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -208,6 +210,120 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+
+        for doc in docs:
+            # make a shallow copy so that adapters can be removed
+            adapters = list(self.table_adapters_by_domain[domain])
```
did you mean to make the copy inside the docs for loop? That seems to defeat the purpose of removing the table in the first place.

snopoke commented on corehq/apps/userreports/tests/test_pillow.py:
```diff
@@ -293,7 +450,7 @@ def test_process_doc_from_sql_stale(self):
             since = self.pillow.get_change_feed().get_latest_offsets()
             form, cases = self._post_case_blocks(i)
-            with self.assertNumQueries(12):
+            with self.assertNumQueries(11):
```
why did this change?

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,33 +103,125 @@ def run(self):
             self._record_checkpoint_in_datadog()
 
+    @property
+    def batch_processors(self):
```
should memoize these to avoid having to re-generate for every change

snopoke commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -100,33 +103,125 @@ def run(self):
+                return
+            retry_changes = set()
+            timer = TimingContext()
+            with timer:
```
This is timing per processor but we're recording the metrics for the pillow as a whole. I do think it's useful to have it be per processor so we should add a tag to the datadog metric so we can separate it out

snopoke reviewed (commented):
I don't think it's a good idea to mix serial processors with batch processors. I think that just complicates things (especially checkpointing). If `processor_chunk_size > 0` then you should validate that all processors support batching.

If we do that then I think it simplifies a few things.

One other point, since we're now including multiple processors in a pillow we should change the datadog metrics to be tagged with the processor name and not the pillow name. @emord do you agree?

sravfeyn commented:
@snopoke sorry, I had few more changes that I haven't pushed that would address some of the points above. Will respond on relevent ones after pushing the changes

sravfeyn commented:
>I don't think it's a good idea to mix serial processors with batch processors. I think that just complicates things (especially checkpointing). If processor_chunk_size > 0 then you should validate that all processors support batching.

If we add that restriction, we won't be able to combine processors without implementing bulk support for all processors. Agree it does complicate things, but I think I have a working version now. Mainly it complicates checkpointing and timer related datadog metrics. Here's what I implemented
1. For checkpointing, if there is a batch processor for the pillow, checkpoint will be updated for the last change in the chunk
2. For metrics, I have consolidated the timers so that the processing time will be the total time to process a set of changes for all processors divided by number of changes. I have updated the tags to add processor.

There's also one important todo - we need to update pillow retry strategy to consider that errors might be on a specific processor instead of all processors, which I am planning to after this PR.

sravfeyn committed:
consolidate checkpoint, timer. Expand support for mix of serial/batch processors

emord commented:
> One other point, since we're now including multiple processors in a pillow we should change the datadog metrics to be tagged with the processor name and not the pillow name. @emord do you agree?

I think that this might depend on the metric? For example a processing time metric could be tagged with processor name, but the checkpoint would still be on the pillow level so it should stay pillow name

> If we add that restriction, we won't be able to combine processors without implementing bulk support for all processors

We don't need to implement "smart" bulk support for the processors. As long as there's an appropriate interface, I think that the following change wouldn't be difficult to add to the base class:

```
def process_changes_bulk(changes):
   for change in changes:
      process_change(change)
```

sravfeyn commented:
> We don't need to implement "smart" bulk support for the processors. As long as there's an appropriate interface, I think that the following change wouldn't be difficult to add to the base class:

I think currently I have a working version without the restriction of all or none batch processors. Checkpoint is updated at pillow level, datadog at processor level

sravfeyn committed:
merge master

sravfeyn committed:
handle table-change error differently; handle individual change exceptions

sravfeyn committed:
terminology

sravfeyn committed:
datadog per processor level

sravfeyn committed:
track success in chunked batch

sravfeyn committed:
Track batch processor exception in dd

sravfeyn committed:
memoize

sravfeyn committed:
simplify

emord commented:
Looks like the test failures are related. I think it's fine to test on staging once you fix those.

This has gotten pretty large. any chance you can go ahead and merge some of the refactoring (like the additions to the adapters and test changes)?

sravfeyn committed:
query count

sravfeyn committed:
fix tests

sravfeyn committed:
account for kwarg

sravfeyn commented:
@emord I took a look and think there aren't big enough changes to push through a separate PR. 

Build passed, so I am gonna test this on staging.

sravfeyn committed:
chunk size 10

sravfeyn committed:
skip reporting on offsets

sravfeyn commented:
Hey guys @emord @snopoke 

I have tested this PR on staging and everything is okay. [Here](https://app.datadoghq.com/notebook/50273/UCR%20BulkPillow%20metrics)'s I have compiled a datadog notebook where you can observe the effect of this PR.

sravfeyn committed:
Merge branch 'master' into sr-batch-ucrpillow

sravfeyn committed:
method update

sravfeyn committed:
group_id to client_id

sravfeyn committed:
use simple client

sravfeyn committed:
no need to refresh

sravfeyn commented:
@emord @snopoke Bumping. Made couple of fixes (due to changes from kafka-settings PR)

sravfeyn commented:
> what do we need the chunk_size tag for?
I thought it might be useful to compare metrics when we need to tune optimum chunk size.

snopoke commented:
Just looking at the data already in datadog it seems like there are large number of different chunk sizes. If there were only one or two I'd be ok with having it but tagging doesn't really work well when you get beyond a certain number of tags.

emord commented on corehq/apps/userreports/pillow.py:
```diff
@@ -272,7 +381,8 @@ class ConfigurableReportKafkaPillow(ConstructedPillow):
 
-    def __init__(self, processor, pillow_name, topics, num_processes, process_num, retry_errors=False):
+    def __init__(self, processor, pillow_name, topics, num_processes, process_num, retry_errors=False,
+            processor_chunk_size=0):
```
why is the default = 0? What does 0 mean in this context?

emord reviewed (commented):
hesitant to merge such a big change when most of the team is about to be offline for a week

sravfeyn committed:
Review

sravfeyn committed:
lint

sravfeyn commented:
> hesitant to merge such a big change when most of the team is about to be offline for a week

Fine to leave it unmerged then, I will ping back post the summit.

snopoke committed:
Merge branch 'master' into sr-batch-ucrpillow
